panel/widgets/file_selector.py

```diff
@@ -0,0 +1,154 @@
+        self._directory = TextInput(value=self.directory, width_policy='max')
+        self._home = Button(name='🏠', width=25, margin=(5, 15, 0, 10), disabled=True)
+        self._back = Button(name='◀', width=25, margin=(5, 10), disabled=True)
+        self._forward = Button(name='▶', width=25, margin=(5, 10), disabled=True)
+        self._up = Button(name='▲', width=25, margin=(5, 10), disabled=True)
+        self._buttons = [self._home, self._back, self._forward, self._up]
+        self._composite = Column(
+            Row(*self._buttons, self._directory, self._go, margin=(0, 10), width_policy='max'),
+            Divider(margin=(0, 20)), self._selector, **layout
+        )
+
+        # Set up state
+        self._stack = []
+        self._cwd = None
+        self._position = -1
+        self._update_files(True)
+
+        # Set up callback
+        self.link(self._directory, directory='value')
+        self._selector.param.watch(self._update_value, 'value')
+        self._go.on_click(self._update_files)
+        self._home.on_click(self._go_home)
+        self._up.on_click(self._go_up)
+        self._back.on_click(self._go_back)
+        self._forward.on_click(self._go_forward)
+        self._directory.param.watch(self._dir_change, 'value')
+        self._selector._lists[False].param.watch(self._select, 'value')
+
+    def _update_value(self, event):
+        value = [v for v in event.new if not self.only_files or os.path.isfile(v)]
+        self._selector.value = value
+        self.value = value
+
+    def _dir_change(self, event):
+        path = os.path.abspath(os.path.expanduser(self._directory.value))
+        if not path.startswith(self.directory):
+            self._directory.value = self.directory
+            return
+        elif path != self._directory.value:
+            self._directory.value = path
+        self._go.disabled = path == self._cwd
+
+    def _update_files(self, event=None):
+        path = os.path.abspath(self._directory.value)
+        if not os.path.isdir(path):
+            self._selector.options = ['Entered path is not valid']
+            self._selector.disabled = True
+            return
+        elif event is not None and (not self._stack or path != self._stack[-1]):
+            self._stack.append(path)
+            self._position += 1
+
+        self._cwd = path
+        self._go.disabled = True
+        self._home.disabled = path == self.directory
+        self._up.disabled = path == self.directory
+        if self._position == len(self._stack)-1:
+            self._forward.disabled = True
+        if 0 <= self._position and len(self._stack) > 1:
+            self._back.disabled = False
+
+        file_paths = glob.glob(os.path.join(path, '*' + self.file_keyword + '*'))
+        files = sorted([p for p in file_paths if os.path.isfile(p)])
+        dir_paths = glob.glob(os.path.join(path, '*'))
+        dirs = sorted([p for p in dir_paths if os.path.isdir(p)])
```

Comment:
Does this filter out links?

Reply:
Good point, I had made a note to think about links but never did test it.